src/ray/core_worker/core_worker.cc

```diff
@@ -728,32 +749,31 @@ bool CoreWorker::AddActorHandle(std::unique_ptr<ActorHandle> actor_handle) {
     // Register a callback to handle actor notifications.
     auto actor_notification_callback = [this](const ActorID &actor_id,
                                               const gcs::ActorTableData &actor_data) {
-      if (actor_data.state() == gcs::ActorTableData::RECONSTRUCTING) {
-        auto it = actor_handles_.find(actor_id);
-        RAY_CHECK(it != actor_handles_.end());
-        if (it->second->IsDirectCallActor()) {
-          // We have to reset the actor handle since the next instance of the
-          // actor will not have the last sequence number that we sent.
-          // TODO: Remove the check for direct calls. We do not reset for the
-          // raylet codepath because it tries to replay all tasks since the
-          // last actor checkpoint.
-          it->second->Reset();
-        }
-      } else if (actor_data.state() == gcs::ActorTableData::DEAD) {
+      RAY_CHECK(actor_data.state() != gcs::ActorTableData::RECONSTRUCTING);
+      if (actor_data.state() == gcs::ActorTableData::DEAD) {
+        direct_actor_submitter_->DisconnectActor(actor_id, true);
```

Comment:
It's not quite clear when will the actor dead state is published to GCS? It is limited to the actor owner or anyone who sends tasks to the actor?  I think with the actor creation task is moved to use direct call, raylet is not able to do this anymore.

If only the owner can publish the dead status, then it seems like we have to learn the connection breakdown without sending the tasks to the actor.

Reply:
Yeah, I think this is a little half baked at the moment (see issue https://github.com/ray-project/ray/issues/6422)

For now, this should be fine since calls to the actor will still fail, it will just take longer.

Reply:
@ericl The tricky part is that the owner might not call the actor after it created the actor (or at least within a certain period), while other callers which actually call the actor can learn the actor failure, they are not authorized to publish the actor death event to gcs. Thus it seems not feasible to rely on ray call from the owner to detect it.

Also some of the scenarios actually requires very prompt actor failure detection (they require the actor being reconstructed within a second).

For now I plan to move the publishing of actor event to raylet in this PR. We can revisit this later and see how this is bestly done. What do you think?

Reply:
Let's do that in another PR. This one is blocking work so we should try to make it mergable first.

Reply:
Hmm, actually now that I understand the current strategy this seems like a good idea to do now. How about this

1. Actor creation is published via the owner when the creation task returns.
2. The node manager that holds the worker for the actor is responsible for marking the actor as dead if the worker disconnects (or if the worker does not exist).

Reply:
Sounds good.

Just wondering what's the benefit of letting the owner to publish the actor creation event?  Is it something related to appending to the redis log?

Reply:
Mostly we want to remove the old scheduler, which this lets us do. The owner publishing the event is also the simplest since it is the one creating it now.

Reply:
Thinking about it a bit more, it seems like there's a problem. If the actor is successfully created, but the owner dies before publishing the actor creation event, then there could be a problem (assuming all the actors are detached and thus could have different lifetime from their owner). If this is done by raylet, since when raylet dies all the workers will die so there would be no issue.

Would you mind explain a bit more on how is this (let owner publish actor creation) related to removal of old scheduler?

Reply:
Well the biggest issue is that there is no longer lineage. Without this, an actor creation task is no longer guaranteed to succeed (or fail) since it's creation task won't be resubmitted, so calls can hang forever.

Moving actor creation to being managed by the caller solves this problem in the same way we handle task failure for direct tasks.

I agree the owner dying causing leaked actors is an issue (see the other issue I linked), but we can solve that separately. This PR is needed otherwise we cannot handle even normal child actor failures.

Reply:
Detached actors will be managed by gcs service anyways right?

Reply:
Thanks for the details, Eric. That helps me understand the issue better.

It seems like which one (raylet or actor owner) to publish actor creation event is not directly related to that actor owner is managing the actors. e.g. Actor owner can still be responsible for actor reconstructions even though actor death is published by raylet.

That said, I agree with you that this could be addressed separately, and I will prioritize on finishing this pr.

Reply:
Would it work if we reuse existing `FinishAssignedActorTask` in node manager to handle direct actor creation task? If that works we don't need a separate gcs table for direct actors.

Reply:
It seems some of the logic there are needed - e.g. `AssignActorId`，`MarkDetachedActor`， which could affect how this worker is going to be handled. I think we should reuse FinishAssignedActorTask if possible?

Reply:
I think we should try keep it separate as much as possible, since all the old code path will be deleted soon. It is risky also to use the existing code since it has different assumptions about lineage, etc.